x/incentive/keeper/rewards.go

```diff
@@ -108,19 +184,202 @@ func (k Keeper) SynchronizeReward(ctx sdk.Context, cdp cdptypes.CDP) {
+	}
+
+	k.SetHardLiquidityProviderClaim(ctx, claim)
+}
+
+// InitializeHardBorrowReward initializes the borrow-side of a hard liquidity provider claim
+// by creating the claim and setting the borrow reward factor index
+func (k Keeper) InitializeHardBorrowReward(ctx sdk.Context, borrow hardtypes.Borrow) {
+	claim, claimFound := k.GetHardLiquidityProviderClaim(ctx, borrow.Borrower)
+	if !claimFound {
+		return
+	}
+
+	var borrowRewardIndexes types.RewardIndexes
+	for _, coin := range borrow.Amount {
+		_, found := k.GetHardBorrowRewardPeriod(ctx, coin.Denom)
+		if !found {
+			continue
+		}
+
+		borrowFactor, foundBorrowFactor := k.GetHardBorrowRewardFactor(ctx, coin.Denom)
+		if !foundBorrowFactor {
+			borrowFactor = sdk.ZeroDec()
+		}
+
+		borrowRewardIndexes = append(borrowRewardIndexes, types.NewRewardIndex(coin.Denom, borrowFactor))
+	}
+
+	claim.BorrowRewardIndexes = borrowRewardIndexes
+
+	k.SetHardLiquidityProviderClaim(ctx, claim)
+}
+
+// SynchronizeHardBorrowReward updates the claim object by adding any accumulated rewards
+// and updating the reward index value
+func (k Keeper) SynchronizeHardBorrowReward(ctx sdk.Context, borrow hardtypes.Borrow) {
+	claim, found := k.GetHardLiquidityProviderClaim(ctx, borrow.Borrower)
+	if !found {
+		return
+	}
+
+	for _, coin := range borrow.Amount {
+		borrowFactor, found := k.GetHardBorrowRewardFactor(ctx, coin.Denom)
+		if !found {
+			continue
+		}
+
+		borrowIndex, BorrowRewardIndex := claim.HasBorrowRewardIndex(coin.Denom)
+		if !BorrowRewardIndex {
+			continue
+		}
+
+		userRewardFactor := claim.BorrowRewardIndexes[borrowIndex].RewardFactor
+		rewardsAccumulatedFactor := borrowFactor.Sub(userRewardFactor)
+		if rewardsAccumulatedFactor.IsZero() {
+			continue
+		}
+		claim.BorrowRewardIndexes[borrowIndex].RewardFactor = borrowFactor
+
+		newRewardsAmount := rewardsAccumulatedFactor.Mul(borrow.Amount.AmountOf(coin.Denom).ToDec()).RoundInt()
+		if newRewardsAmount.IsZero() {
+			continue
+		}
+
+		newRewardsCoin := sdk.NewCoin(types.HardLiquidityRewardDenom, newRewardsAmount)
+		claim.Reward = claim.Reward.Add(newRewardsCoin)
+	}
+
+	k.SetHardLiquidityProviderClaim(ctx, claim)
+}
+
+// UpdateHardSupplyIndexDenoms adds any new deposit denoms to the claim's supply reward index
+func (k Keeper) UpdateHardSupplyIndexDenoms(ctx sdk.Context, deposit hardtypes.Deposit) {
+	claim, found := k.GetHardLiquidityProviderClaim(ctx, deposit.Depositor)
+	if !found {
+		return
+	}
+
+	supplyRewardIndexes := claim.SupplyRewardIndexes
+	for _, coin := range deposit.Amount {
+		_, hasIndex := claim.HasSupplyRewardIndex(coin.Denom)
+		if !hasIndex {
+			supplyFactor, foundSupplyFactor := k.GetHardSupplyRewardFactor(ctx, coin.Denom)
+			if !foundSupplyFactor {
+				supplyRewardIndexes = append(supplyRewardIndexes, types.NewRewardIndex(coin.Denom, supplyFactor))
+			}
+		}
+	}
+	claim.SupplyRewardIndexes = supplyRewardIndexes
+	k.SetHardLiquidityProviderClaim(ctx, claim)
```

Comment:
In the extreme following edge case:

1. User supplies coinA, no incentives - claim doesn't get created.
2. User supplies coinB, no incentives - claim gets created above, but rewardIndexes will still be nil.

To avoid setting a claim in the store which shouldn't exist

```
if claim.SupplyRewardIndexes == []RewardIndexes{} { // not sure exact syntax, could use reflect.DeepEqual I believe
  return
} 
k.SetHardLiquidityProviderClaim(ctx, claim)
```

Reply:
Went with `if len(supplyRewardIndexes) == 0 { return }` in https://github.com/Kava-Labs/kava/pull/757/commits/cd59a7c5dc25240e2f45e7a9facaa9e1966f5528. Since `supplyRewardIndexes` contains all existing and new supply reward denoms we shouldn't set the claim/update the supply reward index if it doesn't have items. However in the case that there are existing supply reward indexes and no new supply reward index denoms, this solution will effectively set claim.SupplyRewardIndexes to itself.